pkg/services/resource.go

```diff
@@ -127,14 +133,41 @@ func (s *sqlResourceService) Replace(ctx context.Context, resource *api.Resource
 	if eErr != nil {
 		return nil, handleUpdateError("Resource", err)
 	}
+
 	return updated, nil
 }
 
-func (s *sqlResourceService) Delete(ctx context.Context, id string) *errors.ServiceError {
-	if err := s.resourceDao.Delete(ctx, id); err != nil {
-		return handleDeleteError("Resource", errors.GeneralError("Unable to delete resource: %s", err))
+func (s *sqlResourceService) UpdateStatus(ctx context.Context, resource *api.Resource) (*api.Resource, *errors.ServiceError) {
+	// Updates the resource status only when its status changes.
+	// If there are multiple requests at the same time, it will cause the race conditions among these
+	// requests (read–modify–write), the advisory lock is used here to prevent the race conditions.
+	lockOwnerID, err := s.lockFactory.NewAdvisoryLock(ctx, resource.ID, db.Resources)
+	if err != nil {
+		return nil, errors.DatabaseAdvisoryLock(err)
 	}
+	defer s.lockFactory.Unlock(ctx, lockOwnerID)
 
+	found, err := s.resourceDao.Get(ctx, resource.ID)
+	if err != nil {
+		return nil, handleGetError("Resource", "id", resource.ID, err)
+	}
+
```

Comment:
I think we may also need to check version here to avoid update a stale resource

Reply:
Nice catch, added version check before update status.